colexec: replace select in linear scan with binary search

helenmhe-zz opened the pull request:
This commit replaces the `SELECT IN` linear scan with a binary search.

Resolves: #40604

helenmhe committed:
colexec: replace SELECT IN linear scan with binary search

Resolves: #40604

Release note (performance improvement): SELECT IN performance on longer filter
rows was improved in the vectorized engine by replacing a linear scan with binary
search.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50337)
<!-- Reviewable:end -->

helenmhe commented:
Bench results on gceworker:

`name                                          old time/op    new time/op    delta
SelectInInt64/useSel=true,hasNulls=true-24      23.7µs ± 1%    13.2µs ± 1%  -44.17%  (p=0.000 n=10+9)
SelectInInt64/useSel=true,hasNulls=false-24     21.5µs ± 0%    11.8µs ± 1%  -44.81%  (p=0.000 n=10+10)
SelectInInt64/useSel=false,hasNulls=true-24     22.9µs ± 1%    13.0µs ± 1%  -43.14%  (p=0.000 n=10+9)
SelectInInt64/useSel=false,hasNulls=false-24    19.7µs ± 0%    10.2µs ± 0%  -48.15%  (p=0.000 n=9+10)

name                                          old speed      new speed      delta
SelectInInt64/useSel=true,hasNulls=true-24     346MB/s ± 1%   465MB/s ± 1%  +34.51%  (p=0.000 n=10+9)
SelectInInt64/useSel=true,hasNulls=false-24    382MB/s ± 0%   519MB/s ± 1%  +36.08%  (p=0.000 n=10+10)
SelectInInt64/useSel=false,hasNulls=true-24    358MB/s ± 1%   473MB/s ± 1%  +32.07%  (p=0.000 n=10+9)
SelectInInt64/useSel=false,hasNulls=false-24   416MB/s ± 0%   603MB/s ± 0%  +44.83%  (p=0.000 n=9+10)`

asubiotto reviewed (changes requested):
Nice improvements!

I would probably squash all of these commits into one because they form one logical change.

Also, by convention, the PR title and commit messages use the package name where most of the changes occur (in this case `colexec`). This doc might be helpful: https://wiki.crdb.io/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages

Both the PR description and the commits require a release note. For this it would be something like `Release note (performance improvement): <describe to users>`

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 1 of 1 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-:-MA1f_fk0VxhRMk87ioM:b-iq748y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @helenmhe)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2-400EwptihHKb31s:-MA2-400EwptihHKb31t:b-gby6jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7db7415a13bc8d40a4044346eb48ecd9e0310b/pkg/sql/colexec/select_in_tmpl.go#L192)):*
> ```Go
> 			hi = i - 1
> 		}
> 		if lo > hi || lo < 0 || hi > len(filterRow) - 1 {
> ```

nit: I think it would be nicer to make this check the for loop exit condition and do the `hasNulls` check after the loop
___
*[pkg/sql/colexec/execgen/cmd/execgen/select_in_gen.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2-_3x9PkNY0KBQFvJ:-MA2-_3x9PkNY0KBQFvK:b-mmhzm3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7db7415a13bc8d40a4044346eb48ecd9e0310b/pkg/sql/colexec/execgen/cmd/execgen/select_in_gen.go#L45)):*
> ```Go
> 	}
> 
> 	tmplInfos := populateTwoArgsOverloads(
> ```

Might be useful to extract this into `overloads_cmp.go` rather than defining these overloads in the gen file. Maybe you can put it right after `sameTypeComparisonOpToOverloads` gets populated in `populateCmpOverloads`. Maybe @yuzefovich has a better idea of where this could go.


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2-400EwptihHKb31s:-MA2HeUF4VggNexvVBfC:b-qsseby) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7db7415a13bc8d40a4044346eb48ecd9e0310b/pkg/sql/colexec/select_in_tmpl.go#L192)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: I think it would be nicer to make this check the for loop exit condition and do the `hasNulls` check after the loop
</blockquote></details>

fixed!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 179 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2SoVHAGhbtCSJwZxf:-MA2SoVHAGhbtCSJwZxg:bmlffzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f6258f6b91103cfcf10f310c7f05efcf455141/pkg/sql/colexec/select_in_tmpl.go#L179)):*
> ```Go
> ) comparisonResult {
> 	lo := 0
> 	hi := len(filterRow) - 1
> ```

The binary search relies on the assumption that elements in `filterRow` are sorted, are they? I think we need to change `fillDatumRow_TYPE` method to do the sort of `result`.

This is another thing that we should check with a unit test.
___
*[pkg/sql/colexec/select_in_tmpl.go, line 184 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2S5jA4t7I7dQ31b1W:-MA2S5jA4t7I7dQ31b1X:b-rzizi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f6258f6b91103cfcf10f310c7f05efcf455141/pkg/sql/colexec/select_in_tmpl.go#L184)):*
> ```Go
> 		var cmpResult int
> 		_ASSIGN_CMP(cmpResult, targetElem, filterRow[i], _, targetCol, _)
> 		switch cmpResult {
> ```

We shouldn't be hardcoding `-1` and `1` values in the case statement because the comparison contract is that
```
// The targetElem will be negative, zero, or positive depending on whether
// leftElem is less-than, equal-to, or greater-than rightElem.
```
(comment from `overloads_base.go`), so we should use `if`s instead.

I think it would also be nice to add a unit test that might have caught this problem (I'd try with `types.Bytes`).
___
*[pkg/sql/colexec/execgen/cmd/execgen/select_in_gen.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2-_3x9PkNY0KBQFvJ:-MA2Q2Lo2-u9QRtZTb1D:b-f4s4yv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7db7415a13bc8d40a4044346eb48ecd9e0310b/pkg/sql/colexec/execgen/cmd/execgen/select_in_gen.go#L45)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Might be useful to extract this into `overloads_cmp.go` rather than defining these overloads in the gen file. Maybe you can put it right after `sameTypeComparisonOpToOverloads` gets populated in `populateCmpOverloads`. Maybe @yuzefovich has a better idea of where this could go.
</blockquote></details>

I wonder why we need this custom call - we already have the same function defined, with the only difference is that we need to call `Compare` on the overload, instead of `Assign`. Take a look at `vec_comparators_tmpl` and `vec_comparators_gen`. Then we can use `sameTypeComparisonOpToOverloads[tree.EQ]` to use with `Execute`.


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @helenmhe, and @yuzefovich)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 179 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2SoVHAGhbtCSJwZxf:-MA2sa1k6-dMBl9ss1f3:b-82pizt) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f6258f6b91103cfcf10f310c7f05efcf455141/pkg/sql/colexec/select_in_tmpl.go#L179)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

The binary search relies on the assumption that elements in `filterRow` are sorted, are they? I think we need to change `fillDatumRow_TYPE` method to do the sort of `result`.

This is another thing that we should check with a unit test.
</blockquote></details>

Resolved because optimizer sorts when normalizing
___
*[pkg/sql/colexec/execgen/cmd/execgen/select_in_gen.go, line 45 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2-_3x9PkNY0KBQFvJ:-MA2sRyu4rg6wPAbvFPq:b-dtkds3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7db7415a13bc8d40a4044346eb48ecd9e0310b/pkg/sql/colexec/execgen/cmd/execgen/select_in_gen.go#L45)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I wonder why we need this custom call - we already have the same function defined, with the only difference is that we need to call `Compare` on the overload, instead of `Assign`. Take a look at `vec_comparators_tmpl` and `vec_comparators_gen`. Then we can use `sameTypeComparisonOpToOverloads[tree.EQ]` to use with `Execute`.
</blockquote></details>

Done, thanks!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_test.go, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA331TYCLvTSsHKb47B:-MA331TYCLvTSsHKb47C:b-oipvin) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L110)):*
> ```Go
> 	for i := 0; i < coldata.BatchSize(); i++ {
> 		if float64(i) < float64(coldata.BatchSize())*selectivity {
> 			col1[i] = -1
> ```

nit: the two cases should be switched (i.e. we should assign `-1` in the `else` block and vice versa).
___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MA31zv67k0yOrB4i2XV:bbe33p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
> ```Go
> 			col1[i] = -1
> 		} else {
> 			col1[i] = 10
> ```

I think it'd be more interesting to randomize this value in `[1, 10]` range for different `i`. Then we will likely to see even bigger perf difference from the old code because in the linear scan value `1` would be found first because it is first in `filterRow`.
___
*[pkg/sql/colexec/select_in_tmpl.go, line 179 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2SoVHAGhbtCSJwZxf:-MA2z-6mDkxlfm21NmNw:b-szcqkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f6258f6b91103cfcf10f310c7f05efcf455141/pkg/sql/colexec/select_in_tmpl.go#L179)):*
<details><summary><i>Previously, helenmhe (Helen He) wrote…</i></summary><blockquote>

Resolved because optimizer sorts when normalizing
</blockquote></details>

Yeah, I didn't realize that the tuple is normalized, nvm.
___
*[pkg/sql/colexec/select_in_tmpl.go, line 184 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2S5jA4t7I7dQ31b1W:-MA2z81H6TaN8seJufrI:b-mjzx3z) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f6258f6b91103cfcf10f310c7f05efcf455141/pkg/sql/colexec/select_in_tmpl.go#L184)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

We shouldn't be hardcoding `-1` and `1` values in the case statement because the comparison contract is that
```
// The targetElem will be negative, zero, or positive depending on whether
// leftElem is less-than, equal-to, or greater-than rightElem.
```
(comment from `overloads_base.go`), so we should use `if`s instead.

I think it would also be nice to add a unit test that might have caught this problem (I'd try with `types.Bytes`).
</blockquote></details>

Discussed it on slack. TL;DR is that it seems like we currently actually return `-1/0/1` when comparing all types, but it is not enforced in any way and the comments suggest that we are free to return `negative/0/positive` instead. Helen ran a quick benchmark to see the performance difference between using `switch` and `if` which came out to 4-7% slower for the latter, and I think we should use `if` given the absolute speeds are very high anyway.

Disregard comment about adding a unit test that would catch the problem (but adding a few test cases would be beneficial).
___
*[pkg/sql/colexec/select_in_tmpl.go, line 179 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA30qnY1r2SI7apYJwK:-MA30qnY1r2SI7apYJwL:b-5ev8cs) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_tmpl.go#L179)):*
> ```Go
> ) comparisonResult {
> 	lo := 0
> 	hi := len(filterRow) - 1
> ```

I wonder whether you chose this particular implementation of binary search based on the benchmarks. I think more idiomatic way for `hi` index to be exclusive (i.e. operate on `[lo, hi)` range during the search).
___
*[pkg/sql/colexec/select_in_tmpl.go, line 193 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA32UhP0DAvcf_RYd9X:-MA32UhP0DAvcf_RYd9Y:b53y5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_tmpl.go#L193)):*
> ```Go
> 		}
> 
> 		if lo > hi || lo < 0 || hi > len(filterRow)-1 {
> ```

I think we might be able to eliminate a condition or two in here. If we switch to searching in `[lo, hi)` range, then we'd be able to `break` simply when `lo == hi`.

nit: I'd move the conditions into `for` condition.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
nit: PR title should be renamed to be similar to our guidelines on the commit messages.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-:-MA39orA6bDBikuhVdbz:bvf0jlu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 179 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2SoVHAGhbtCSJwZxf:-MA39kaz1LRR3Fk25pSZ:blj2cha) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f6258f6b91103cfcf10f310c7f05efcf455141/pkg/sql/colexec/select_in_tmpl.go#L179)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yeah, I didn't realize that the tuple is normalized, nvm.
</blockquote></details>

We should add this observation as a comment in the code.


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @helenmhe, and @yuzefovich)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 179 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA2SoVHAGhbtCSJwZxf:-MA2sr5tDBqMGJ2TUVSC:b-8b11j9) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f6258f6b91103cfcf10f310c7f05efcf455141/pkg/sql/colexec/select_in_tmpl.go#L179)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

We should add this observation as a comment in the code.
</blockquote></details>

Added
___
*[pkg/sql/colexec/select_in_tmpl.go, line 193 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA32UhP0DAvcf_RYd9X:-MA3IO2e9BOgLgKkxiYs:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_tmpl.go#L193)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think we might be able to eliminate a condition or two in here. If we switch to searching in `[lo, hi)` range, then we'd be able to `break` simply when `lo == hi`.

nit: I'd move the conditions into `for` condition.
</blockquote></details>

Done!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 175 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA3O6UFB8WyIKZ4ipiq:-MA3O6UFB8WyIKZ4ipir:bazs59d) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e8590e46eb5ec9f9cbddac89790cfab18f5b6e9/pkg/sql/colexec/select_in_tmpl.go#L175)):*
> ```Go
> }
> 
> // Filter row input is already sorted due to normalization
> ```

nit: missing period. Also, I'd move the comment into the function itself and expand it (saying something like "Filter row input ..., so we can use the binary search right away.")
___
*[pkg/sql/colexec/select_in_tmpl.go, line 190 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA3ORQcAb_rHnVQAjDE:-MA3ORQcAb_rHnVQAjDF:b-ti08ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e8590e46eb5ec9f9cbddac89790cfab18f5b6e9/pkg/sql/colexec/select_in_tmpl.go#L190)):*
> ```Go
> 			lo = i + 1
> 		} else {
> 			hi = i - 1
> ```

I think this needs to be `hi = i` since we've changed `hi` to be exclusive (because `i-1` element might be the one we're searching for). It might be interesting to come up with a test case for this.


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @helenmhe, and @yuzefovich)

___
*[pkg/sql/colexec/select_in_tmpl.go, line 190 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA3ORQcAb_rHnVQAjDE:-MA3PJgF2SIPvMc5xEIT:bmea8ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e8590e46eb5ec9f9cbddac89790cfab18f5b6e9/pkg/sql/colexec/select_in_tmpl.go#L190)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this needs to be `hi = i` since we've changed `hi` to be exclusive (because `i-1` element might be the one we're searching for). It might be interesting to come up with a test case for this.
</blockquote></details>

Whoops you're right, thanks!


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @helenmhe, and @yuzefovich)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MA3PbqNAHoBCYqR4uUc:b-i1ap3k) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think it'd be more interesting to randomize this value in `[1, 10]` range for different `i`. Then we will likely to see even bigger perf difference from the old code because in the linear scan value `1` would be found first because it is first in `filterRow`.
</blockquote></details>

Yeah I was originally thinking that but not sure how to handle randomization without creating too much variability in the benchmark?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MA3QAGIB3iYcLfWfxcH:bntgnee) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, helenmhe (Helen He) wrote…</i></summary><blockquote>

Yeah I was originally thinking that but not sure how to handle randomization without creating too much variability in the benchmark?
</blockquote></details>

I think some variability is acceptable. The important thing is that we measure runtime on the whole batch (so on 1024 rows), so if we choose a distribution something like:
```
  50% - not present (-1)
  5% - for each value in 1..10 range
```
then regardless of a particular order in which these values appear, the speed should be more or less stable on the whole batch.

True, if we randomly get value `1` in 10% of cases and value `10` in 0%, the speed will probably be noticeably different in a single run, but that will be hidden by the `go bench` tooling as noise.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MA3R7Gb0ZASdNrVkeQI:b2mieu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think some variability is acceptable. The important thing is that we measure runtime on the whole batch (so on 1024 rows), so if we choose a distribution something like:
```
  50% - not present (-1)
  5% - for each value in 1..10 range
```
then regardless of a particular order in which these values appear, the speed should be more or less stable on the whole batch.

True, if we randomly get value `1` in 10% of cases and value `10` in 0%, the speed will probably be noticeably different in a single run, but that will be hidden by the `go bench` tooling as noise.
</blockquote></details>

(I guess 1 and 10 won't be that different, but 6 and 10 will be because we will land on 6 in the very first try of the binary search.)


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> Reviewed 1 of 1 files at r5, 2 of 2 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @helenmhe and @yuzefovich)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MA6dT6w5w5KxFgp_vWk:b-t6mdye) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

(I guess 1 and 10 won't be that different, but 6 and 10 will be because we will land on 6 in the very first try of the binary search.)
</blockquote></details>

I agree with the desire but disagree with the method. If we would like to benchmark different cases, I think those should be explicit defined cases rather than randomly generated. This also gives us the nice property of running a specific scenario without having to comment code out, which is tedious.
___
*[pkg/sql/colexec/select_in_tmpl.go, line 178 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA6eIBHB6UptbbBbHKW:-MA6eIBHB6UptbbBbHKX:bb9uf3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/321e457d4605a187c6170c691af9e90fab3b41c3/pkg/sql/colexec/select_in_tmpl.go#L178)):*
> ```Go
> 	targetElem _GOTYPE, targetCol _GOTYPESLICE, filterRow []_GOTYPE, hasNulls bool,
> ) comparisonResult {
> 	// Filter row input is already sorted due to normalization, so we can use a
> ```

@raduberinde can we unconditionally assume that the `IN` clause is sorted?


<!-- Sent from Reviewable.io -->

asubiotto commented:
@helenmhe it makes it easier to review the PR if you squash all your commits so that we can review only one version of the code (most of the time, sometimes multiple commits are justified). Throughout the review/update process, there usually isn't a change to the commit message since it's a general description of the change. This post is a helpful resource on squashing: https://www.internalpointers.com/post/squash-commits-into-one-git. Let me know if you need help!

asubiotto reviewed (approved):
Sorry for the delay, I didn't realize this was ready for another review! Feel free to write "Friendly ping" or "RFAL" in the PR if you feel like some time goes by without a review.

In the release note, could you mention that this performance improvement is specific to the vectorized engine?

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-:-MAWeUzZ4wC8mSqcjV5A:b-f1u4mm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @helenmhe and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MAWpttu1xcljDJlM77t:b-1qgfx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I agree with the desire but disagree with the method. If we would like to benchmark different cases, I think those should be explicit defined cases rather than randomly generated. This also gives us the nice property of running a specific scenario without having to comment code out, which is tedious.
</blockquote></details>

What is the disadvantage of having a distribution of values as I described above? Currently, we're always searching for value 10 (which is the last one in the `IN` tuple) or value -1 (which is not present), so if we were solely relying on such benchmark, we would "overfit" for this particular scenario.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @helenmhe and @yuzefovich)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MAWquRy38Xv1PLJWZtX:bevx8xn) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

What is the disadvantage of having a distribution of values as I described above? Currently, we're always searching for value 10 (which is the last one in the `IN` tuple) or value -1 (which is not present), so if we were solely relying on such benchmark, we would "overfit" for this particular scenario.
</blockquote></details>

The disadvantage is adding randomization to benchmarks which makes comparing across branches hard unless you manually change the code. I think we should avoid doing this wherever we can. I agree that adding more scenarios is good though. Maybe we should add a search for `1` to give the advantage to the linear search.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MAWs4XXBDdH7e1ciXLh:b-arxscx) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

The disadvantage is adding randomization to benchmarks which makes comparing across branches hard unless you manually change the code. I think we should avoid doing this wherever we can. I agree that adding more scenarios is good though. Maybe we should add a search for `1` to give the advantage to the linear search.
</blockquote></details>

So is your argument that if we change the data distribution now in the benchmark, it will be apples-vs-oranges comparison when comparing against the old branch (e.g. release-20.1)?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @helenmhe and @yuzefovich)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MAWtn1z6ZLYzlrTLzFg:b-5jvjci) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

So is your argument that if we change the data distribution now in the benchmark, it will be apples-vs-oranges comparison when comparing against the old branch (e.g. release-20.1)?
</blockquote></details>

No. My argument is that once this benchmark is introduced, any future changes will have a hard time comparing themselves to this new version of the code (with the random benchmark). If one run randomizes the constant choice to be 5 and the other 10, the benchmarks aren't comparable.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MAX7rSb6Piupro1Rrh-:bjyrxh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

No. My argument is that once this benchmark is introduced, any future changes will have a hard time comparing themselves to this new version of the code (with the random benchmark). If one run randomizes the constant choice to be 5 and the other 10, the benchmarks aren't comparable.
</blockquote></details>

I agree with this. At the same time I would like for this argument to not block the improvements to the benchmark itself, so here is a simple solution: we create a separate first commit that modifies the benchmark to the way we want it to be, and the second commit modifies the algorithm, then we can run the benchmarks of 1st commit vs 2nd commit to get the improvement of the change to the algorithm. Thoughts?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @helenmhe and @yuzefovich)

___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MAXR8hsC9IFf39i96U8:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I agree with this. At the same time I would like for this argument to not block the improvements to the benchmark itself, so here is a simple solution: we create a separate first commit that modifies the benchmark to the way we want it to be, and the second commit modifies the algorithm, then we can run the benchmarks of 1st commit vs 2nd commit to get the improvement of the change to the algorithm. Thoughts?
</blockquote></details>

SGTM


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50337)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @helenmhe)

___
*[pkg/sql/colexec/select_in_test.go, line 110 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA331TYCLvTSsHKb47B:-MAg2JXD93jozxzacnDF:b-ecqdjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L110)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: the two cases should be switched (i.e. we should assign `-1` in the `else` block and vice versa).
</blockquote></details>

I didn't mean to switch the sign as well - I think that "philosophically correct" pattern is
```
if float64(i) < float64(coldata.BatchSize())*selectivity {
  col1[i] = 10
} else {
  col1[i] = -1
}
```
because "selectivity" means "how many tuples are selected" - i.e. they are "present" in the batch. Originally the meaning was switched, and it just happened to work ok because `selectivity` is set to `0.5`.
___
*[pkg/sql/colexec/select_in_test.go, line 112 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/50337#-MA31zv67k0yOrB4i2XU:-MAg30kG0LqZhzTfmyX8:b-nhoqpe) ([raw file](https://github.com/cockroachdb/cockroach/blob/90bd3e64d3b563d54d0238a7fb9e177514be4634/pkg/sql/colexec/select_in_test.go#L112)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

SGTM
</blockquote></details>

Helen, as per this discussion, could you add another commit (that would be first in this PR) that improves the benchmark itself, and then the binary search change would be the second commit?


<!-- Sent from Reviewable.io -->

asubiotto commented:
Friendly ping @helenmhe. It would be good to get this PR in so we can close the loop on the SELECT IN improvement.

helenmhe committed:
colexec: expand input distribution for SELECT IN benchmark

The input distribution was previously 1 or -1 with equal probability for a filter
row of {1, 2, 3}, which was optimal for a linear scan but did not capture the long
filter row worst-case well. This commit changes the filter row to be between 0 and
9, and replaces the 1 input with a near-uniform distribution between 0 and 9.

Release note: None

asubiotto commented:
LGTM. I think CI is complaining that you need to run `make generate` to ensure the generated code is up to date now that we check the generated code in.

helenmhe commented:
TFTRs!

bors r+
